src/tir/schedule/primitive/blockize_tensorize.cc

```diff
@@ -0,0 +1,638 @@
+   *   - Check the sub AST is equal to the description function.
+   *
+   * Mutate:
+   *   - Blockize the sub AST (please refer blockize for details)
+   *   - Bind buffers
+   *   - Mutate the implementation of the tensor intrinsic by replacing its buffers with new
+   *     buffers created via match buffer region.
+   *   - Replace the sub tree with the mutated function.
+   */
+  const auto* loop = loop_sref->StmtAs<ForNode>();
+  CHECK(loop) << "Only support tensorize a loop for now";
+
+  const auto* desc_block_realize =
+      Downcast<BlockRealize>(intrinsic->description->body)->block->body.as<BlockRealizeNode>();
+  const Block& desc_block = desc_block_realize->block;
+  const auto* impl_block_realize =
+      Downcast<BlockRealize>(intrinsic->implementation->body)->block->body.as<BlockRealizeNode>();
+  Block impl_block = impl_block_realize->block;
+
+  // Step 1: Blockize the subtree rooted at the given loop
+  const StmtSRef& block_sref = Blockize(self, loop_sref);
+  const BlockRealize& block_realize = GetBlockRealize(self, block_sref);
+
+  // Step 2: Compare the block with the description of the tensor intrinsic, find the correspondence
+  // between buffers in the block and the description.
+  TensorizeComparator comparator(/*assert_mode=*/true);
+  comparator.VisitStmt(block_realize, GetRef<Stmt>(desc_block_realize));
+
+  // Step 3: Find the correspondence between buffers in the current AST and the implementation of
+  // the tensor intrinsic
+
+  // Step 3.1: Map from intrinsic func buffer to description func buffer
+  std::unordered_map<Buffer, Buffer, ObjectPtrHash, ObjectPtrEqual> intrin_buffer_map;
+  RemapTensorIntrinBuffers(intrinsic, &intrin_buffer_map);
+  // Step 3.2: Map form intrinsic func buffer to current AST buffer
+  std::unordered_map<Buffer, Buffer, ObjectPtrHash, ObjectPtrEqual> buffer_map;
+  for (const auto& pair : intrin_buffer_map) {
+    auto it = comparator.rhs_buffer_map_.find(pair.second);
+    ICHECK(it != comparator.rhs_buffer_map_.end());
+    buffer_map[pair.first] = it->second;
+  }
+
+  // Step 4: Create MatchBufferRegion for the params of the implementation function of the tensor
+  // intrin to make them subregions of the buffer in the original IR.
+  std::unordered_map<Buffer, Array<Range>, ObjectPtrHash, ObjectPtrEqual> buffer_region_map;
+  for (const auto& read : impl_block->reads) {
+    buffer_region_map.emplace(read->buffer, read->region);
+  }
+  for (const auto& write : impl_block->writes) {
+    buffer_region_map.emplace(write->buffer, write->region);
+  }
+  Array<MatchBufferRegion> match_buffer_regions;
+  for (size_t i = 0; i < intrinsic->implementation->params.size(); ++i) {
+    const auto& param = intrinsic->implementation->params[i];
+    const auto& buffer = intrinsic->implementation->buffer_map.at(param);
+    const auto& source = buffer_map.at(buffer);
+    Region region = buffer_region_map.at(buffer);
+    auto extra_indices = comparator.buffer_indices_.at(source);
```

Comment:
nit: i found it a little bit confusing to use auto here, so...

```suggestion
    const Var& param = intrinsic->implementation->params[i];
    Buffer buffer = intrinsic->implementation->buffer_map.at(param);
    Buffer source = buffer_map.at(buffer);
    Array<Range> region = buffer_region_map.at(buffer);
    const std::vector<PrimExpr>& extra_indices = comparator.buffer_indices_.at(source);
```